File: lingting-spring-grpc/src/main/java/live/lingting/spring/grpc/configuration/GrpcServerConfiguration.java
package live.lingting.spring.grpc.configuration;

import io.grpc.BindableService;
import io.grpc.ServerInterceptor;
import live.lingting.framework.grpc.GrpcServerBuilder;
import live.lingting.framework.grpc.interceptor.GrpcServerTraceIdInterceptor;
import live.lingting.framework.grpc.properties.GrpcServerProperties;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;

import java.util.List;

/**
 * @author lingting 2024-02-05 16:23
 */
public class GrpcServerConfiguration {

	@Bean
	@ConditionalOnMissingBean
	public GrpcServerBuilder grpcServerBuilder(GrpcServerProperties properties, List<BindableService> services,
			List<ServerInterceptor> interceptors) {
		return new GrpcServerBuilder().properties(properties).service(services).interceptor(interceptors);
	}

	@Bean
	@ConditionalOnMissingBean
	public GrpcServerTraceIdInterceptor grpcServerTraceIdInterceptor(GrpcServerProperties properties) {
		return new GrpcServerTraceIdInterceptor(properties);
	}

}

File: lingting-spring-grpc/src/main/kotlin/live/lingting/spring/grpc/configuration/GrpcServerConfiguration.kt
package live.lingting.spring.grpc.configuration

import io.grpc.BindableService
import io.grpc.ServerInterceptor
import live.lingting.framework.grpc.GrpcServerBuilder
import live.lingting.framework.grpc.customizer.ServerCustomizer
import live.lingting.framework.grpc.interceptor.GrpcServerTraceIdInterceptor
import live.lingting.framework.grpc.properties.GrpcServerProperties
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Bean

/**
 * @author lingting 2024-02-05 16:23
 */
open class GrpcServerConfiguration {

    @Bean
    @ConditionalOnMissingBean
    fun grpcServerBuilder(
        properties: GrpcServerProperties,
        services: Collection<BindableService>,
        interceptors: Collection<ServerInterceptor>,
        customizers: Collection<ServerCustomizer>
    ): GrpcServerBuilder {
        return GrpcServerBuilder(customizers).properties(properties).service(services).interceptor(interceptors)
    }

    @Bean
    @ConditionalOnMissingBean
    fun grpcServerTraceIdInterceptor(properties: GrpcServerProperties): GrpcServerTraceIdInterceptor {
        return GrpcServerTraceIdInterceptor(properties)
    }

}
